src/components/views/rooms/wysiwyg_composer/hooks/useSuggestion.ts

```diff
@@ -0,0 +1,196 @@
+ * into the composer
+ * - `onSelect`: a selection change listener to be attached to the plain text composer
+ * - `suggestion`: if the cursor is inside something that could be interpreted as a command or a mention,
+ * this will be an object representing that command or mention, otherwise it is null
+ *
+ */
+
+export function useSuggestion(
+    editorRef: React.RefObject<HTMLDivElement>,
+    setText: (text: string) => void,
+): {
+    handleMention: (link: string, text: string, attributes: Attributes) => void;
+    handleCommand: (text: string) => void;
+    onSelect: (event: SyntheticEvent<HTMLDivElement>) => void;
+    suggestion: MappedSuggestion | null;
+} {
+    const [suggestion, setSuggestion] = useState<Suggestion>(null);
+
+    // TODO handle the mentions (@user, #room etc)
+    const handleMention = (): void => {};
+
+    // We create a `seletionchange` handler here because we need to know when the user has moved the cursor,
+    // we can not depend on input events only
+    const onSelect = (): void => processSelectionChange(editorRef, suggestion, setSuggestion);
+
+    const handleCommand = (replacementText: string): void =>
+        processCommand(replacementText, editorRef, suggestion, setSuggestion, setText);
+
+    return {
+        suggestion: mapSuggestion(suggestion),
+        handleCommand,
+        handleMention,
+        onSelect,
+    };
+}
+
+/**
+ * Convert a PlainTextSuggestionPattern (or null) to a MappedSuggestion (or null)
+ *
+ * @param suggestion - the suggestion that is the JS equivalent of the rust model's representation
+ * @returns - null if the input is null, a MappedSuggestion if the input is non-null
+ *
+ */
+export const mapSuggestion = (suggestion: Suggestion): MappedSuggestion | null => {
+    if (suggestion === null) {
+        return null;
+    } else {
+        const { node, startOffset, endOffset, ...mappedSuggestion } = suggestion;
+        return mappedSuggestion;
+    }
+};
+
+/**
+ * Replaces the relevant part of the editor text with the replacement text after a command is selected
+ * from the autocomplete.
+ *
+ * @param replacementText - the text that we will insert into the DOM
+ * @param editorRef - a ref to the editor
+ * @param suggestion - representation of the part of the DOM that will be replaced
+ * @param setSuggestion - a setter to allow the suggestion state to be updated
+ * @param setText - a setter to update the state of the composer content after replacing a mention
+ * or a command
+ */
+export const processCommand = (
+    replacementText: string,
+    editorRef: React.RefObject<HTMLDivElement>,
+    suggestion: Suggestion,
+    setSuggestion: React.Dispatch<React.SetStateAction<Suggestion>>,
+    setText: (text: string) => void,
+): void => {
+    // if we do not have any of the values we need to do the work, do nothing
+    if (suggestion === null || editorRef.current === null) {
```

Comment:
we don't seem to use `editorRef` elsewhere in this function - do we really need it?

Reply:
Nope - removed it and updated tests accordingly